core/analysis/multi_delimited_token_stream.cpp

```diff
@@ -0,0 +1,538 @@
+      offset.end = std::distance(start_, std::addressof(*next));
+
+      if (next == data_.end()) {
+        data_ = {};
+      } else {
+        data_ = bytes_view{next + skip, data_.end()};
+      }
+
+      return true;
+    }
+  }
+};
+
+template<typename Derived>
+class MultiDelimitedTokenStreamSingleCharsBase
+  : public MultiDelimitedTokenStreamBase<
+      MultiDelimitedTokenStreamSingleCharsBase<Derived>> {
+ public:
+  auto FindNextDelim() {
+    auto where = static_cast<Derived*>(this)->FindNextDelim();
+    return std::make_pair(where, 1);
+  }
+};
+
+template<std::size_t N>
+class MultiDelimitedTokenStreamSingleChars final
+  : public MultiDelimitedTokenStreamSingleCharsBase<
+      MultiDelimitedTokenStreamSingleChars<N>> {
+ public:
+  explicit MultiDelimitedTokenStreamSingleChars(
+    const MultiDelimitedAnalyser::Options& opts) {
+    IRS_ASSERT(opts.delimiters.size() == N);
+    std::size_t k = 0;
+    for (const auto& delim : opts.delimiters) {
+      IRS_ASSERT(delim.size() == 1);
+      bytes[k++] = delim[0];
+    }
+  }
+
+  auto FindNextDelim() {
+    return std::search(this->data_.begin(), this->data_.end(), bytes.begin(),
+                       bytes.end());
+  }
+
+  std::array<byte_type, N> bytes;
+};
+
+template<>
+class MultiDelimitedTokenStreamSingleChars<1> final
+  : public MultiDelimitedTokenStreamSingleCharsBase<
+      MultiDelimitedTokenStreamSingleChars<1>> {
+ public:
+  explicit MultiDelimitedTokenStreamSingleChars(
+    const MultiDelimitedAnalyser::Options& opts) {
+    IRS_ASSERT(opts.delimiters.size() == 1);
+    IRS_ASSERT(opts.delimiters[0].size() == 1);
+    delim = opts.delimiters[0][0];
+  }
+
+  auto FindNextDelim() {
+    if (auto pos = this->data_.find(delim); pos != bstring::npos) {
+      return this->data_.begin() + pos;
+    }
+    return this->data_.end();
+  }
+
+  byte_type delim;
+};
+
+template<>
+class MultiDelimitedTokenStreamSingleChars<0> final
+  : public MultiDelimitedTokenStreamSingleCharsBase<
+      MultiDelimitedTokenStreamSingleChars<0>> {
+ public:
+  explicit MultiDelimitedTokenStreamSingleChars(
+    const MultiDelimitedAnalyser::Options& opts) {
+    IRS_ASSERT(opts.delimiters.empty());
+  }
+
+  auto FindNextDelim() { return this->data_.end(); }
+};
+
+class MultiDelimitedTokenStreamGenericSingleChars final
+  : public MultiDelimitedTokenStreamSingleCharsBase<
+      MultiDelimitedTokenStreamGenericSingleChars> {
+ public:
+  explicit MultiDelimitedTokenStreamGenericSingleChars(const Options& opts) {
+    for (const auto& delim : opts.delimiters) {
+      IRS_ASSERT(delim.size() == 1);
+      bytes_[delim[0]] = true;
+    }
+  }
+
+  auto FindNextDelim() {
+    return std::find_if(data_.begin(), data_.end(), [&](auto c) {
+      if (c > SCHAR_MAX) {
+        return false;
+      }
+      IRS_ASSERT(c <= SCHAR_MAX);
+      return bytes_[c];
+    });
+  }
+  // TODO maybe use a bitset instead?
+  std::array<bool, SCHAR_MAX + 1> bytes_;
```

Comment:
```suggestion
  std::array<bool, SCHAR_MAX + 1> bytes_{};
```
otherwise, it would contain garbage